### ISSUE TITLE
Security Issue - Unrestricted Access to Sensitive Files on the Server

### ISSUE THREAD

**Issue by Cyber-Guy1** [commenthol/md-fileserver#17]
Dear Team,

I would like to bring to your attention a security issue that has been identified in our server. There is a lack of restrictions on the server's URI and accessible files. As a result, it is possible for unauthorized users to access sensitive internal files. For example, if a user enters the command "mdstart list.md" in their terminal and specifies "[http://localhost:4000/../../../../../../../../../../../../../../..//etc/passwd](http://localhost:4000//etc/passwd)" in their browser, the contents of the "passwd" file will be displayed. This vulnerability can be exploited by attackers to gain access to sensitive files from users who are actively using the service.

Upon analysis, it appears that the root cause of this problem can be traced back to the following code snippet:

```
  const token = appConfig.token()
  const filename = path.resolve(process.cwd(), cmd.file || '')
  const url = `http://${config.hostname}:${cmd.port}` +
    filename2uri(filename) + `?session=${token}`

  if (cmd.output) {
    const _config = Object.assign({}, config, appConfig.config)
    render(filename, '', _config)
      .then(data => fs.writeFileSync(cmd.output, data))
      .catch(console.error)

    return
  }
```

The issue arises due to the lack of validation of the file name, ensuring it matches the one entered by the user in the command prompt. As a result, the "render" function accepts the file name directly, without verifying if it has been modified or if it has moved to a different scope that is beyond the intended boundaries. 

To remediate the security issue in the provided code snippet, we need to perform proper validation of the file path and restrict access to sensitive files. Here's an updated version of the code:

```
#!/usr/bin/env node

const fs = require('fs')
const path = require('path')
const child = require('child_process')
const isPortOpen = require('../lib/checkport')
const config = require('../config')
const { filename2uri } = require('../lib/utils')
const appConfig = require('../lib/appConfig')()
const { render } = require('../lib/markd')

const validateFilePath = (filePath) => {
  const normalizedPath = path.normalize(filePath)
  const resolvedPath = path.resolve(normalizedPath)
  const basePath = path.resolve(process.cwd())

  // Ensure that the resolved path is within the allowed basePath
  return resolvedPath.startsWith(basePath)
}

const argv = () => {
  // ... (unchanged code)
}

const cmd = argv()

isPortOpen(cmd, (isOpen) => {
  const startBrowser = cmd.file && cmd.browser

  const token = appConfig.token()
  const filename = cmd.file || ''
  const filePath = path.resolve(process.cwd(), filename)

  if (!validateFilePath(filePath)) {
    console.error('Invalid file path or access to sensitive files is restricted.')
    return
  }

  const url = `http://${config.hostname}:${cmd.port}` +
    filename2uri(filename) + `?session=${token}`

  // ... (rest of the code unchanged)
})

// ... (unchanged code)
```

The validateFilePath function has been added to ensure the resolved path is within the allowed base path (the current working directory). This helps in preventing unauthorized access to sensitive files.

## Proof-of-Concept:

![image](https://user-images.githubusercontent.com/66295316/226573028-6a046d55-c739-4833-8bd8-86ce1b90dfc4.png)

**Comment by commenthol** [comment]
﻿Hi @cyber-Guy1,

Unfortunately I have to reject your finding, as the file exposed is only
accessible by the user who started the server. 

The server only starts locally, that means that only "another" user logged into
the same machine would be able to access the opened port. See commit
037ad8bb48ff696f6430b453aa9e9962f7d5cc77. 

In order to prevent that such "another" user would be able to access files from
the "primary" user, a random session token is generated and set as session
cookie. Without that token the "another" user will only get a 401 Unauthorized.
This was fixed in commit 20edd0bf007333e225585710e280fc2737861586. 

Hint: Please check the Session-Cookie
in your Postman-Request. This shows that you are the "primary" user having
started the server.

The server can only expose files the "primary" user and only that user has
already read access. Therefore your finding is not exploiting any information to
the very same user.

If you are able to see the /etc/passwd file as "another" user, then please
consider filing a security incident.

**Comment by Cyber-Guy1** [comment]
Dear @commenthol,

Thank you for your response. I understand your concerns regarding the security finding that I reported, and I appreciate the information you have provided about the server's configuration.

However, I would like to respectfully disagree with your decision to reject my finding. While it may be true that the file exposed is only accessible by the user who started the server, the session token that is being sent through GET URL can still be vulnerable to referer leaks, MiTM attacks, and other security threats. Additionally, the lack of sufficient expiry for the session token can leave the server exposed to potential breaches.

In light of these concerns, I strongly recommend that you take the necessary steps to remediate the issue by implementing the suggested code changes that I have provided in my original report. By doing so, you can ensure that the server is secure against potential attacks and that the confidentiality and integrity of user data are preserved.

Thank you for your attention to this matter, and please let me know if you have any questions or concerns.

Sincerely,
